### PR TITLE
Wrapped errors

### PR DESCRIPTION
**Pull request by iulianpascalau**
- added custom wrapped errors that can automatically store the line that called the Wrap method. Interchangeable with the standard `error` interface

### CODE REVIEW

**Review by matei-p — DISMISSED** [review]

File: core/errors/wrappedErrors.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+// WrapWithError wraps the target error with the provided one
+func (werr *wrappableError) WrapWithError(err error) WrappableError {
+	return werr.wrapWithErrorWithSkipLevels(err, skipLevels)
+}
+
+// GetBaseError gets the core error
+func (werr *wrappableError) GetBaseError() error {
+	errorSlice := werr.errsWithLocation
+	return errorSlice[0].err
```
Comment (bogdan-rosianu) [inline core/errors/wrappedErrors.go:50]: not sure if accessing slice without checking it's length is the best approach, but can remaing like this. `errsWithLocation` cannot be an empty slice

File: core/errors/wrappedErrors.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+		errWithLocation := errorSlice[len(errorSlice)-1-idxErr]
+		errMsg := nilPlaceholder
+		if errWithLocation.err != nil {
+			errMsg = errWithLocation.err.Error()
+		}
+		suffix := ""
+		if errMsg != "" {
+			suffix = " [" + errMsg + "]"
+		}
+		strErr += "\n\t" + errWithLocation.location + suffix
```
Comment (bogdan-rosianu) [inline core/errors/wrappedErrors.go:93]: the entire error will begin with a blank space. please refactor

File: core/errors/wrappedErrors_test.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	assert.NotNil(t, we.GetLastError())
+	assert.True(t, strings.Contains(we.Error(), "/core/errors/wrappedErrors_test.go"))
+	assert.True(t, strings.Contains(we.Error(), nilPlaceholder))
+
+	we = WrapError(nil)
+	msg := "a message"
+	we = we.WrapWithMessage(msg)
+	require.NotNil(t, we)
+	assert.NotNil(t, we.GetLastError())
+	assert.True(t, strings.Contains(we.Error(), "/core/errors/wrappedErrors_test.go"))
```
Comment (bogdan-rosianu) [inline core/errors/wrappedErrors_test.go:54]: you could have assigned the `we.Error()` to a variable, and then use it for checks

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: core/errors/wrappedErrors.go (hunk 2 above)
Comment (matei-p) [reply core/errors/wrappedErrors.go:93]: it looked prettier in console, but might be useless

File: core/errors/wrappedErrors.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+// WrapError constructs a WrappableError from an error
+func WrapError(err error) WrappableError {
+	errAsWrappable, ok := err.(WrappableError)
+	if ok {
+		return errAsWrappable
+	}
+	return &wrappableError{
+		errsWithLocation: []errorWithLocation{createErrorWithLocation(err, 2)},
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply core/errors/wrappedErrors.go:28]: done

File: core/errors/wrappedErrors.go (hunk 1 above)
Comment (iulianpascalau) [reply core/errors/wrappedErrors.go:50]: there is no way we can have a 0 length slice according to the code.

File: core/errors/wrappedErrors.go (hunk 2 above)
Comment (iulianpascalau) [reply core/errors/wrappedErrors.go:93]: done

File: core/errors/wrappedErrors_test.go (hunk 3 above)
Comment (iulianpascalau) [reply core/errors/wrappedErrors_test.go:54]: done and then reverted as it defeats the purpose of easily understanding the tests.

**Review by bogdan-rosianu — APPROVED** [review]

**Review by matei-p — APPROVED** [review]

File: core/errors/wrappedErrors.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
+package errors
+
+import (
+	"errors"
+	"fmt"
+	"runtime"
+)
+
+const skipStackLevels = 2
```
Comment (miiu96) [inline core/errors/wrappedErrors.go:9]: maybe you can group this consts 
```
const (
 const1..
 const2...
)
```

**Review by camilbancioiu — APPROVED** [review]

**Pull request merged**